### PR TITLE
Add support: Atorch S1WP energy monitoring plug with display

### LINKED ISSUE #1103: Request support for Atorch (S1WP) device Smart Plug Energy meter
Could you please add Atorch (S1WP) to the support device list please.

I tried to add it myself but when I add yaml file in adding device I receive Unknown error. when I remove the file the suggested devices to choose are not energy meters!

Thank you in advance!

From Tuya iot API explorer:

    "model": "S1WP",
    "name": "Smart Energy Meter",
    "product_id": "sqrf2g1amfutn4co",
    "product_name": "炬为电能宝-计控型(S1WP)",

{
  "result": {
    "properties": [
      {
        "code": "switch_1",
        "custom_name": "",
        "dp_id": 1,
        "time": 1691582860910,
        "value": true
      },
      {
        "code": "countdown_1",
        "custom_name": "",
        "dp_id": 9,
        "time": 1695000549682,
        "value": 0
      },
      {
        "code": "add_ele",
        "custom_name": "",
        "dp_id": 17,
        "time": 1695003267591,
        "value": 5
      },
      {
        "code": "cur_current",
        "custom_name": "",
        "dp_id": 18,
        "time": 1695005887553,
        "value": 830
      },
      {
        "code": "cur_power",
        "custom_name": "",
        "dp_id": 19,
        "time": 1695005887568,
        "value": 6613
      },
      {
        "code": "cur_voltage",
        "custom_name": "",
        "dp_id": 20,
        "time": 1695005887514,
        "value": 12385
      },
      {
        "code": "price",
        "custom_name": "",
        "dp_id": 101,
        "time": 1695000549682,
        "value": 73
      },
      {
        "code": "cost",
        "custom_name": "",
        "dp_id": 102,
        "time": 1695005872416,
        "value": 51192
      },
      {
        "code": "add_cost",
        "custom_name": "",
        "dp_id": 103,
        "time": 1695004926491,
        "value": 5
      },
      {
        "code": "ovp",
        "custom_name": "",
        "dp_id": 104,
        "time": 1695000549682,
        "value": 2650
      },
      {
        "code": "ocp",
        "custom_name": "",
        "dp_id": 105,
        "time": 1695000549682,
        "value": 1000
      },
      {
        "code": "opp",
        "custom_name": "",
        "dp_id": 106,
        "time": 1695000549682,
        "value": 3000
      },
      {
        "code": "language",
        "custom_name": "",
        "dp_id": 107,
        "time": 1695000549682,
        "value": "english"
      },
      {
        "code": "work_value",
        "custom_name": "",
        "dp_id": 108,
        "time": 1695000549682,
        "value": 9
      },
      {
        "code": "standby_value",
        "custom_name": "",
        "dp_id": 109,
        "time": 1695000549682,
        "value": 3
      },
      {
        "code": "standby_time",
        "custom_name": "",
        "dp_id": 110,
        "time": 1695000549682,
        "value": 60
      },
      {
        "code": "beep",
        "custom_name": "",
        "dp_id": 111,
        "time": 1695000549682,
        "value": true
      },
      {
        "code": "sw_mode",
        "custom_name": "",
        "dp_id": 112,
        "time": 1695000549682,
        "value": "normally_open"
      },
      {
        "code": "data_reset",
        "custom_name": "",
        "dp_id": 113,
        "time": 1689902912733,
        "value": false
      },
      {
        "code": "wifi_reset",
        "custom_name": "",
        "dp_id": 114,
        "time": 1695000475501,
        "value": true
      },
      {
        "code": "factor_reset",
        "custom_name": "",
        "dp_id": 115,
        "time": 1690317870702,
        "value": true
      },
      {
        "code": "screen_rotation",
        "custom_name": "",
        "dp_id": 116,
        "time": 1690317641569,
        "value": true
      },
      {
        "code": "standby_screen",
        "custom_name": "",
        "dp_id": 117,
        "time": 1695000549682,
        "value": "measurement"
      },
      {
        "code": "menu",
        "custom_name": "",
        "dp_id": 118,
        "time": 1695000549682,
        "value": "safety_protection"
      },
      {
        "code": "plow",
        "custom_name": "",
        "dp_id": 119,
        "time": 1695000549682,
        "value": 5
      },
      {
        "code": "time_a",
        "custom_name": "",
        "dp_id": 120,
        "time": 1695000549682,
        "value": 10
      },
      {
        "code": "phigh",
        "custom_name": "",
        "dp_id": 121,
        "time": 1695000549682,
        "value": 500
      },
      {
        "code": "time_b",
        "custom_name": "",
        "dp_id": 122,
        "time": 1695000549682,
        "value": 1
      },
      {
        "code": "ele",
        "custom_name": "",
        "dp_id": 123,
        "time": 1695005872423,
        "value": 70126
      },
      {
        "code": "run_countdown",
        "custom_name": "",
        "dp_id": 124,
        "time": 1695000475516,
        "value": 0
      },
      {
        "code": "timing_close",
        "custom_name": "",
        "dp_id": 125,
        "time": 1695000549682,
        "value": 10800
      },
      {
        "code": "timing_open",
        "custom_name": "",
        "dp_id": 126,
        "time": 1695000549682,
        "value": 60
      },
      {
        "code": "loop_open",
        "custom_name": "",
        "dp_id": 127,
        "time": 1695000549682,
        "value": 60
      },
      {
        "code": "loop_close",
        "custom_name": "",
        "dp_id": 128,
        "time": 1695000549682,
        "value": 60
      },
      {
        "code": "countdown_open",
        "custom_name": "",
        "dp_id": 129,
        "time": 1695000551639,
        "value": 60
      },
      {
        "code": "countdown_close",
        "custom_name": "",
        "dp_id": 130,
        "time": 1695000551639,
        "value": 60
      },
      {
        "code": "r_mode",
        "custom_name": "",
        "dp_id": 131,
        "time": 1695000551639,
        "value": "open"
      },
      {
        "code": "warning",
        "custom_name": "",
        "dp_id": 132,
        "time": 1695000551639,
        "value": "off"
      },
      {
        "code": "cur_frequency",
        "custom_name": "",
        "dp_id": 133,
        "time": 1695005882593,
        "value": 6012
      },
      {
        "code": "power_factor",
        "custom_name": "",
        "dp_id": 134,
        "time": 1695005887574,
        "value": 64
      },
      {
        "code": "cpu_temp",
        "custom_name": "",
        "dp_id": 135,
        "time": 1695005882605,
        "value": 34
      },
      {
        "code": "price_mode",
        "custom_name": "",
        "dp_id": 136,
        "time": 1695000551639,
        "value": "single_rate"
      },
      {
        "code": "over_time",
        "custom_name": "",
        "dp_id": 137,
        "time": 1695000551639,
        "value": 0
      },
      {
        "code": "ttl",
        "custom_name": "",
        "dp_id": 138,
        "time": 1695000551639,
        "value": "memory"
      }
    ]
  },
  "success": true,
  "t": 1695005892345,
  "tid": "3489316d55cf11eeacc91e124c4e4d14"
}

### PR DESCRIPTION
**Pull request by blakadder**
fixes #1103

### CODE REVIEW

File: custom_components/tuya_local/devices/atorch_s1wp.yaml [hunk 1]
```diff
@@ -0,0 +1,575 @@
...
+    dps:
+      - id: 9
+        type: integer
+        name: value
+        unit: min
+        range:
+          min: 0
+          max: 360000
+        mapping:
+          - scale: 60
```
Comment (make-all) [inline custom_components/tuya_local/devices/atorch_s1wp.yaml:26]: Two default mappings here, with different scales.

File: custom_components/tuya_local/devices/atorch_s1wp.yaml [hunk 2]
```diff
@@ -0,0 +1,575 @@
...
+    dps:
+      - id: 101
+        type: integer
+        name: value
+        range:
+          min: 0
+          max: 99999
+        mapping:
+          - scale: 100
+            step: 1
```
Comment (make-all) [inline custom_components/tuya_local/devices/atorch_s1wp.yaml:75]: `step: 1` is not needed.

File: custom_components/tuya_local/devices/atorch_s1wp.yaml [hunk 3]
```diff
@@ -0,0 +1,575 @@
...
+    dps:
+      - id: 106
+        type: integer
+        name: value
+        unit: W
+        range:
+          min: 1
+          max: 4500
+        mapping:
+          - scale: 1
```
Comment (make-all) [inline custom_components/tuya_local/devices/atorch_s1wp.yaml:127]: `scale: 1` is also not needed, so this whole mapping block can be removed.

File: custom_components/tuya_local/devices/atorch_s1wp.yaml [hunk 4]
```diff
@@ -0,0 +1,575 @@
...
+  - entity: button
+    name: Factory reset
+    category: diagnostic
+    icon: "mdi:autorenew"
+    dps:
+      - id: 115
+        type: boolean
+        name: button
+        optional: true
+  - entity: button
```
Comment (make-all) [inline custom_components/tuya_local/devices/atorch_s1wp.yaml:231]: Is button correct? Do you send another "true" to rotate it back?

**Review by make-all — COMMENTED** [review]
I only commented once on the not needed scale and step, but they are repeated throughout.

File: custom_components/tuya_local/devices/atorch_s1wp.yaml (hunk 3 above)
Comment (blakadder) [reply custom_components/tuya_local/devices/atorch_s1wp.yaml]: i was just annoyed by debug log recommending to add mapping for the dp

File: custom_components/tuya_local/devices/atorch_s1wp.yaml (hunk 4 above)
Comment (blakadder) [reply custom_components/tuya_local/devices/atorch_s1wp.yaml:215]: yes, it triggers the action in the plug, it doesn't have states

**Pull request merged**